oauth2client/client.py

```diff
@@ -1617,6 +1617,23 @@ def _revoke(self, http_request):
         """
         self._do_revoke(http_request, self.access_token)
 
+    def sign_blob(self, blob):
+        """Cryptographically sign a blob (of bytes).
+
+        Args:
+            blob: bytes, Message to be signed.
+
+        Returns:
+            tuple, A pair of the private key ID used to sign the blob and
+            the signed contents.
+        """
+        raise NotImplementedError
+
+    @property
+    def service_account_email(self):
+        """Get the email for the current service account."""
+        raise NotImplementedError
```

Comment:
I'm not thrilled about `AssertionCredentials` being a public class, but since it is, we have to consider not just our own library's known subclasses but also all possible subclasses that may be out there that others have implemented. I suspect that the only basis we could ever have for being certain that this property would make sense for all `AppAssertionCredentials` everywhere would come from something about the `AppAssertionCredentials` class itself. But there's nothing here in the class motivating that - it's just coincidence that our three subclasses happen to have this other quality in common. So I think this abstract method should be dropped from this change.

There's one other thing we _could_ do, though: if we believe that there's a fundamental "sameness" to the way our three classes present their service account email and we'd like to present that sameness as part of the library's class taxonomy, we could define a separate purely-virtual `class ServiceAccountEmailHaver` (final name TBD :-P) and have our three classes implement that.